File: php/run-tests.php
<?php

declare(strict_types=1);

/**
 * Test runner script for RocketReach PHP SDK
 * 
 * This script runs all tests and generates coverage reports
 */

require_once __DIR__ . '/vendor/autoload.php';

use PHPUnit\TextUI\Command;

class TestRunner
{
    private const COVERAGE_DIR = __DIR__ . '/coverage';
    private const PHPUNIT_CONFIG = __DIR__ . '/phpunit.xml';

    public function runAllTests(): int
    {
        echo "Running all tests with coverage...\n";
        
        $this->ensureCoverageDirectory();
        
        $command = new Command();
        $args = [
            'phpunit',
            '--configuration=' . self::PHPUNIT_CONFIG,
            '--coverage-html=' . self::COVERAGE_DIR . '/html',
            '--coverage-text=' . self::COVERAGE_DIR . '/coverage.txt',
            '--coverage-clover=' . self::COVERAGE_DIR . '/clover.xml',
            '--testdox'
        ];
        
        return $command->run($args, false);
    }

    public function runUnitTests(): int
    {
        echo "Running unit tests...\n";
        
        $command = new Command();
        $args = [
            'phpunit',
            '--configuration=' . self::PHPUNIT_CONFIG,
            '--testsuite=Unit',
            '--testdox'
        ];
        
        return $command->run($args, false);
    }

    public function runIntegrationTests(): int
    {
        echo "Running integration tests...\n";
        
        $command = new Command();
        $args = [
            'phpunit',
            '--configuration=' . self::PHPUNIT_CONFIG,
            '--testsuite=Integration',
            '--testdox'
        ];
        
        return $command->run($args, false);
    }

    public function runE2ETests(): int
    {
        echo "Running E2E tests...\n";
        
        $command = new Command();
        $args = [
            'phpunit',
            '--configuration=' . self::PHPUNIT_CONFIG,
            '--testsuite=E2E',
            '--testdox'
        ];
        
        return $command->run($args, false);
    }

    public function runLiveTests(): int
    {
        echo "Running live API tests...\n";
        
        $command = new Command();
        $args = [
            'phpunit',
            '--configuration=' . self::PHPUNIT_CONFIG,
            '--testsuite=Live',
            '--testdox'
        ];
        
        return $command->run($args, false);
    }

    public function runTestsUntilGreen(): int
    {
        $maxAttempts = 10;
        $attempt = 1;
        
        while ($attempt <= $maxAttempts) {
            echo "\n=== Test Run Attempt {$attempt} ===\n";
            
            $exitCode = $this->runAllTests();
            
            if ($exitCode === 0) {
                echo "\n✅ All tests passed! Coverage achieved.\n";
                return 0;
            }
            
            echo "\n❌ Tests failed. Attempt {$attempt}/{$maxAttempts}\n";
            $attempt++;
            
            if ($attempt <= $maxAttempts) {
                echo "Retrying in 5 seconds...\n";
                sleep(5);
            }
        }
        
        echo "\n❌ Tests failed after {$maxAttempts} attempts.\n";
        return $exitCode;
    }

    private function ensureCoverageDirectory(): void
    {
        if (!is_dir(self::COVERAGE_DIR)) {
            mkdir(self::COVERAGE_DIR, 0755, true);
        }
    }

    public function showCoverageReport(): void
    {
        $coverageFile = self::COVERAGE_DIR . '/coverage.txt';
        
        if (file_exists($coverageFile)) {
            echo "\n=== Coverage Report ===\n";
            echo file_get_contents($coverageFile);
        } else {
            echo "Coverage report not found. Run tests first.\n";
        }
    }
}

// Command line interface
if (php_sapi_name() === 'cli') {
    $runner = new TestRunner();
    
    $command = $argv[1] ?? 'all';
    
    switch ($command) {
        case 'all':
            $exitCode = $runner->runAllTests();
            $runner->showCoverageReport();
            exit($exitCode);
            
        case 'unit':
            exit($runner->runUnitTests());
            
        case 'integration':
            exit($runner->runIntegrationTests());
            
        case 'e2e':
            exit($runner->runE2ETests());
            
        case 'live':
            exit($runner->runLiveTests());
            
        case 'loop':
            exit($runner->runTestsUntilGreen());
            
        case 'coverage':
            $runner->showCoverageReport();
            exit(0);
            
        default:
            echo "Usage: php run-tests.php [all|unit|integration|e2e|live|loop|coverage]\n";
            echo "  all         - Run all tests with coverage\n";
            echo "  unit        - Run unit tests only\n";
            echo "  integration - Run integration tests only\n";
            echo "  e2e         - Run E2E tests only\n";
            echo "  live        - Run live API tests only\n";
            echo "  loop        - Run tests until 100% green\n";
            echo "  coverage    - Show coverage report\n";
            exit(1);
    }
}

File: python/run_tests.py
#!/usr/bin/env python3
"""
Test Runner

Comprehensive test runner for the RocketReach Python SDK.
"""

import sys
import os
import subprocess
import time
from typing import List, Dict, Any

# Add the src directory to the Python path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), 'src'))


class TestRunner:
    """Test runner for the RocketReach Python SDK."""
    
    def __init__(self):
        self.test_categories = [
            "unit",
            "integration", 
            "e2e",
            "live"
        ]
        self.results = {}
        self.total_tests = 0
        self.passed_tests = 0
        self.failed_tests = 0
    
    def run_tests(self, category: str = None, verbose: bool = True) -> Dict[str, Any]:
        """
        Run tests for a specific category or all categories.
        
        Args:
            category: Test category to run (unit, integration, e2e, live)
            verbose: Whether to show verbose output
            
        Returns:
            Dict containing test results
        """
        if category:
            categories = [category]
        else:
            categories = self.test_categories
        
        print("=" * 60)
        print("RocketReach Python SDK Test Runner")
        print("=" * 60)
        print()
        
        for cat in categories:
            print(f"Running {cat} tests...")
            result = self._run_category_tests(cat, verbose)
            self.results[cat] = result
            
            if result['success']:
                print(f"✅ {cat} tests passed: {result['passed']}/{result['total']}")
            else:
                print(f"❌ {cat} tests failed: {result['failed']}/{result['total']}")
            
            self.total_tests += result['total']
            self.passed_tests += result['passed']
            self.failed_tests += result['failed']
            print()
        
        self._print_summary()
        return self.results
    
    def _run_category_tests(self, category: str, verbose: bool) -> Dict[str, Any]:
        """Run tests for a specific category."""
        test_dir = f"tests/{category}"
        
        if not os.path.exists(test_dir):
            return {
                'success': True,
                'total': 0,
                'passed': 0,
                'failed': 0,
                'message': f"No {category} tests found"
            }
        
        cmd = [
            sys.executable, "-m", "pytest",
            test_dir,
            "-v" if verbose else "",
            "--tb=short",
            "--cov=rocketreach" if category == "unit" else "",
            "--cov-report=term-missing" if category == "unit" else "",
            "--cov-fail-under=100" if category == "unit" else "",
        ]
        
        # Remove empty strings
        cmd = [arg for arg in cmd if arg]
        
        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=300  # 5 minute timeout
            )
            
            # Parse pytest output to get test counts
            passed = 0
            failed = 0
            total = 0
            
            if result.returncode == 0:
                # Count passed tests
                for line in result.stdout.split('\n'):
                    if ' PASSED ' in line:
                        passed += 1
                    elif ' FAILED ' in line:
                        failed += 1
                total = passed + failed
                success = failed == 0
            else:
                # Count failed tests
                for line in result.stdout.split('\n'):
                    if ' FAILED ' in line:
                        failed += 1
                    elif ' PASSED ' in line:
                        passed += 1
                total = passed + failed
                success = False
            
            return {
                'success': success,
                'total': total,
                'passed': passed,
                'failed': failed,
                'returncode': result.returncode,
                'stdout': result.stdout,
                'stderr': result.stderr
            }
            
        except subprocess.TimeoutExpired:
            return {
                'success': False,
                'total': 0,
                'passed': 0,
                'failed': 0,
                'message': f"{category} tests timed out"
            }
        except Exception as e:
            return {
                'success': False,
                'total': 0,
                'passed': 0,
                'failed': 0,
                'message': f"Error running {category} tests: {str(e)}"
            }
    
    def _print_summary(self):
        """Print test summary."""
        print("=" * 60)
        print("TEST SUMMARY")
        print("=" * 60)
        print(f"Total Tests: {self.total_tests}")
        print(f"Passed: {self.passed_tests}")
        print(f"Failed: {self.failed_tests}")
        
        if self.total_tests > 0:
            success_rate = (self.passed_tests / self.total_tests) * 100
            print(f"Success Rate: {success_rate:.1f}%")
        
        print()
        
        if self.failed_tests == 0:
            print("🎉 ALL TESTS PASSED!")
            print("✅ The Python SDK is ready for production use.")
        else:
            print("⚠️  Some tests failed. Please review the output above.")
        
        print("=" * 60)
    
    def run_loop(self, max_iterations: int = 10) -> bool:
        """
        Run tests in a loop until all pass or max iterations reached.
        
        Args:
            max_iterations: Maximum number of iterations
            
        Returns:
            True if all tests pass, False otherwise
        """
        print("Starting test loop...")
        print(f"Maximum iterations: {max_iterations}")
        print()
        
        for iteration in range(1, max_iterations + 1):
            print(f"Iteration {iteration}/{max_iterations}")
            print("-" * 40)
            
            results = self.run_tests(verbose=False)
            
            if self.failed_tests == 0:
                print(f"\n🎉 All tests passed after {iteration} iteration(s)!")
                return True
            
            print(f"\n⚠️  {self.failed_tests} tests failed. Retrying...")
            print()
            
            # Reset counters for next iteration
            self.total_tests = 0
            self.passed_tests = 0
            self.failed_tests = 0
        
        print(f"\n❌ Tests failed after {max_iterations} iterations.")
        return False


def main():
    """Main function."""
    import argparse
    
    parser = argparse.ArgumentParser(description="RocketReach Python SDK Test Runner")
    parser.add_argument("--category", choices=["unit", "integration", "e2e", "live"], 
                       help="Test category to run")
    parser.add_argument("--loop", action="store_true", 
                       help="Run tests in a loop until all pass")
    parser.add_argument("--max-iterations", type=int, default=10,
                       help="Maximum iterations for loop mode")
    parser.add_argument("--verbose", action="store_true",
                       help="Verbose output")
    
    args = parser.parse_args()
    
    runner = TestRunner()
    
    if args.loop:
        success = runner.run_loop(args.max_iterations)
        sys.exit(0 if success else 1)
    else:
        results = runner.run_tests(args.category, args.verbose)
        
        # Exit with error code if any tests failed
        if runner.failed_tests > 0:
            sys.exit(1)
        else:
            sys.exit(0)


if __name__ == "__main__":
    main()
